### PR TITLE
[EigenSolversApp] Refactoring

### PR DESCRIPTION
**Pull request by oberbichler**
Some major refactoring of the `EigenSolversApplication` to make the implementation of new features easier. The templating stuff is reduced. Each solver has it's own header. Now it is possible to read settings from the Parameters and use solver-specific features.

### CODE REVIEW

File: applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h [hunk 1]
```diff
@@ -166,12 +114,12 @@ class EigenDirectSolver
...
-        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > x(rX.data().begin(), rX.size());
-        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > b(rB.data().begin(), rB.size());
+        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1>> x(rX.data().begin(), rX.size());
+        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1>> b(rB.data().begin(), rB.size());
 
-        x = m_solver.solve(b);
+        const bool success = m_solver.Solve(b, x);
 
-        KRATOS_ERROR_IF(m_solver.info() != Eigen::Success) << "Solving failed!" << std::endl;
+        KRATOS_ERROR_IF(!success) << "Decomposition failed!" << std::endl;
```
Comment (armingeiser) [inline applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h]: This is only Solution right? decomposition happened in InitializeSolutionStep with m_solver.Compute(a)

File: applications/EigenSolversApplication/custom_solvers/eigen_pardiso_ldlt_solver.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+template <typename TScalar = double>
+class EigenPardisoLDLTSolver
+{
+public:
+    using Scalar = TScalar;
+    using SparseMatrix = Eigen::SparseMatrix<Scalar, Eigen::RowMajor, int>;
+    using Vector = Eigen::Matrix<Scalar, Eigen::Dynamic, 1>;
+
+private:
+    Eigen::PardisoLDLT<SparseMatrix> m_solver;
```
Comment (armingeiser) [inline applications/EigenSolversApplication/custom_solvers/eigen_pardiso_ldlt_solver.h:37]: as far as i understand there are two layers now?
* EigenDirectSolver
  * EigenPardisoLDLTSolver
    * Eigen::PardisoLDLT
and the childs are always called m_solver?

File: applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    static std::string Name()
+    {
+        return "eigen_sparse_cg";
+    }
+
+    void Initialize(Parameters settings)
+    {
+    }
+
+    bool Compute(Eigen::Map<const SparseMatrix> a)
```
Comment (armingeiser) [inline applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h:49]: Havent used Eigen in a while, should Map be passed as reference?

**Review by armingeiser — CHANGES_REQUESTED** [review]
Few minor comments, but it looks much easier to customize now, thanks!

File: applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h (hunk 1 above)
Comment (oberbichler) [reply applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h]: 👍 copy&paste mistake...

File: applications/EigenSolversApplication/custom_solvers/eigen_pardiso_ldlt_solver.h (hunk 2 above)
Comment (oberbichler) [reply applications/EigenSolversApplication/custom_solvers/eigen_pardiso_ldlt_solver.h:37]: It is the same as before. The last "layer" was just hidden by the templating

File: applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h (hunk 3 above)
Comment (oberbichler) [reply applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h:49]: Since `Map` is a reference I would say that there is no overhead.

Comment (armingeiser) [reply applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h:49]: I was expecting that, was just not sure anymore.

**Review by armingeiser — DISMISSED** [review]

**Comment by oberbichler** [comment]
should i worry about the gitlab check?

**Comment by philbucher** [comment]
> should i worry about the gitlab check?

nope, it is not fully functional yet
Would you mind waiting until I had time to take a closer look?
I will do it asap
Thx

**Comment by oberbichler** [comment]
@philbucher The output should now also be available for the eigenvalue solver.

File: applications/EigenSolversApplication/custom_solvers/eigen_sparse_qr_solver.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        x = m_solver.solve(b);
+
+        const bool success = m_solver.info() == Eigen::Success;
+
+        return success;
+    }
+
+    void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << "EigenDirectSolver <" << Name() << "> finished.";
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_sparse_qr_solver.h:69]:
```suggestion
        rOStream << "EigenDirectSolver <" << Name() << ">";
```

File: applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        x = m_solver.solve(b);
+
+        const bool success = m_solver.info() == Eigen::Success;
+
+        return success;
+    }
+
+    void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << "EigenDirectSolver <" << Name() << "> finished.";
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_sparse_cg_solver.h:71]:
```suggestion
        rOStream << "EigenDirectSolver <" << Name() << ">";
```

File: applications/EigenSolversApplication/custom_solvers/eigen_pardiso_lu_solver.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        x = m_solver.solve(b);
+
+        const bool success = m_solver.info() == Eigen::Success;
+
+        return success;
+    }
+
+    void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << "EigenDirectSolver <" << Name() << "> finished.";
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_pardiso_lu_solver.h:69]:
```suggestion
        rOStream << "EigenDirectSolver <" << Name() << ">";
```

File: applications/EigenSolversApplication/custom_solvers/eigen_pardiso_llt_solver.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        x = m_solver.solve(b);
+
+        const bool success = m_solver.info() == Eigen::Success;
+
+        return success;
+    }
+
+    void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << "EigenDirectSolver <" << Name() << "> finished.";
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_pardiso_llt_solver.h:69]:
```suggestion
        rOStream << "EigenDirectSolver <" << Name() << ">";
```

File: applications/EigenSolversApplication/custom_solvers/eigen_pardiso_ldlt_solver.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        x = m_solver.solve(b);
+
+        const bool success = m_solver.info() == Eigen::Success;
+
+        return success;
+    }
+
+    void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << "EigenDirectSolver <" << Name() << "> finished.";
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_pardiso_ldlt_solver.h:69]:
```suggestion
        rOStream << "EigenDirectSolver <" << Name() << ">";
```

File: applications/EigenSolversApplication/custom_solvers/eigen_sparse_lu_solver.h [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        const bool success = m_solver.info() == Eigen::Success;
+
+        KRATOS_ERROR_IF(!success) << m_solver.lastErrorMessage() << std::endl;
+
+        return success;
+    }
+
+    void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << "EigenDirectSolver <" << Name() << "> finished.";
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_sparse_lu_solver.h:70]:
```suggestion
        rOStream << "EigenDirectSolver <" << Name() << ">";
```

File: applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h [hunk 10]
```diff
@@ -166,12 +114,12 @@ class EigenDirectSolver
...
-        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > x(rX.data().begin(), rX.size());
-        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > b(rB.data().begin(), rB.size());
+        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1>> x(rX.data().begin(), rX.size());
+        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1>> b(rB.data().begin(), rB.size());
 
-        x = m_solver.solve(b);
+        const bool success = m_solver.Solve(b, x);
 
-        KRATOS_ERROR_IF(m_solver.info() != Eigen::Success) << "Solving failed!" << std::endl;
+        KRATOS_ERROR_IF(!success) << "Solving failed!" << std::endl;
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h]: in order to further unify I suggest to introduce a method `PrintAdditionalErrorMessages` (name just a suggestion)
=> those could be overridden by the eigen but not the pardiso solvers to print the messages
```suggestion
        KRATOS_ERROR_IF(!success) << "Solving failed!\n" << GetSolverErrorMessage() << std::endl;
```

**Review by philbucher — COMMENTED** [review]
suggestion, we can discuss it in person

**Review by philbucher — APPROVED** [review]
:+1:

**Comment by philbucher** [comment]
@oberbichler now it does not compile without MKL, can you have a look please?

**Pull request merged**